### ISSUE TITLE
Feature request: load Salome mesh dat file in BEMRosetta

### ISSUE THREAD

**Issue by Ran-Tu** [BEMRosetta/BEMRosetta#19]
Dear developer,

Would it be possible to add up new features that can load Salome mesh dat file in BEMRosetta.
Salome is an open-source software which can create geometry and mesh. Therefore, I think it can be a good preprocessor for BEM simulation. 

Below is a zipped folder including three files. The file with name OC3Spar.dat is the mesh file from Salome. It is half of a spar foundation. The file with name Dat2Pnl.f90 is a fortran code programmed by myself. I programmed this code to convert Salome data file to HAMS pnl file. And the file with name OC3Spar.pnl is the converted results, which can be visualized in BEMRosetta.
[OC3Spar.zip](https://github.com/BEMRosetta/BEMRosetta/files/10948807/OC3Spar.zip)

Please allow me to introduce this "dat" format based on my understanding.

line 1 includes two integers: 951 and 1084. The first integer is the total number of nodes. The second integer is the total number of elements, including edge elements, surface elements and brick elements. So in my example, it has 951 nodes and 1084 elements. And my example has no brick elements. But there exists edge elements and surface elements.

Line 2 ~ 952 : the totally 951 rows described 951 node information. Each row includes one integer and three real numbers. The integer is the index of this node, followed by x-coordinate, y-coordinate and z-coordinate. For example, line 2 includes: 1 4.70000000000000e+00 0.00000000000000e+00 -1.20000000000000e+02. That denotes the node index is 1. node x-coordinate is 4.70000000000000e+00, node y-coordinate is 0.00000000000000e+00, and node z-coordinate is -1.20000000000000e+02.

Line 953 ~ 2036: the totally 1084 rows described 1084 element information. Each row includes 4 or 5 integers. The first integer is the element index. The second integer denotes the type of this element.  It is hundreds. The hundreds place denotes the dimension of this element( i.e. number 1 denotes edge elements, number 2 denotes surface element and number 3 denotes brick elements). And the ones place denotes the number of node connectivity(i.e number 3 denotes that this element has 3 nodes, so it is triangle element. Number 4 denotes that this element has 4 nodes, so it is quadrilateral element). 
For example, line 953 is: 1 102 5 3. That means the element index is 1, it is a edge element and it has two nodes. The index of these two nodes are 5 and 3 respectively.
Line 1149 is: 197 204 1 41 42 16.  That means the element index is 197, it is a surface element and it has four nodes. The index of these four nodes are 1, 41, 42, 16 respectively. The sequence of node index will determine the face normal direction.
Line 1156 is: 204 203 35 34 48.  That means the element index is 204, it is a surface element and it has three nodes. The index of these three nodes are 35,34,48 respectively. The sequence of node index will determine the face normal direction.

Above is an introduction to Salome dat file format. For BEM, I think we only need surface element with type 203 and 204.

As I have mentioned, above zipped folder has a fortran code to convert Salome dat file to HAMS pnl file. The key idea of this code is to extract all nodes and all surface elements with type 203 and 204, but exclude edge elements. 

Dear developer, if you have interest to implement new features that can load Salome mesh dat file in BEMRosetta, Although this code is not elegant, I hope it can reduce sort of your work.

Regards,

**Comment by Ran-Tu** [comment]
I found that Salome mesh dat file has something in common with Nemoh mesh dat file. With a few changes, BEMRosetta's code reading Nemoh mesh dat file may be adapted to Salome mesh dat file.

**Comment by izabala123** [comment]
Dear Ran.Tu

Thank you for your proposal. I will check it out. I'll let you know soon.

Best regards
Iñaki

**Comment by izabala123** [comment]
Hi Ran.Tu

It has been very easy. Tomorrow you will have the new version ready:
[image: image.png]
Iñaki

**Comment by Ran-Tu** [comment]
Thanks for your work. I am look forwarding to the new version.

**Comment by Ran-Tu** [comment]
Dear @izabala123 ,

I downloaded and used the updated version. I possibly found a bug.
1. I imported a floating platform which totally has 760 nodes and 1421 triangular panels. 
But when I check the summary, it shows 760 nodes and 1420 panels (seen in below figure)
![Screenshot 2023-03-14 182914](https://user-images.githubusercontent.com/56197603/224972694-04a3757d-9080-40d1-8bfa-17eda5600146.png)

2. Then I exported this mesh to HAMS pnl mesh. I found it has 761 nodes and 1420 panel.

That means the export process created a extra node with index 761. And I also found this node was used in panel with index 285 (seen in below figure).

![Screenshot 2023-03-14 184017](https://user-images.githubusercontent.com/56197603/224975583-6377eaf1-467e-4ae2-8637-85a7647b76d0.png)

But actually in this file, all panels should be triangle as I did not create any quadrilateral panel in Salome. So I guess there might be a bug.

The dat file and exported pnl file are uploaded below.
[Group_under.zip](https://github.com/BEMRosetta/BEMRosetta/files/10967331/Group_under.zip)

**Comment by izabala123** [comment]
Thank you Ran-Tu

The problem came that the original mesh is just the positive side of Y axis. When saving it including symmetries, BEMRosetta cuts the mesh by the XZ plane. As some points had infinitesimal negative Y, BEMRosetta cut them and added some new point. However the results would be the same, you can see them in the pictures below, as the new quad had a new point really colinear. Anyway, I have reviewed the functions to cut the mesh, and now a tiny tolerance is included. In your case, the new saved mesh doesn't have added nodes or panels. Please try it and tell me if now it works.

Best regards
Iñaki

![image](https://user-images.githubusercontent.com/38589221/225564662-4e0a2461-2da0-43e2-9af2-241f0a76abb0.png)
![image](https://user-images.githubusercontent.com/38589221/225564754-645b4d23-dc36-4ebe-a8dd-6d0696bef676.png)

**Comment by Ran-Tu** [comment]
Dear @izabala123 

I imported my case in the updated BEMRosetta. But when I wanted to export the HAMS.pnl file, an error raised as follow.
![1678970076905](https://user-images.githubusercontent.com/56197603/225618593-0725bb0b-274e-4b99-9a32-0e02554a424f.png)

**Comment by izabala123** [comment]
Hi Ran

I think it is solved. Please check it.

Best regards
Iñaki

**Comment by Ran-Tu** [comment]
Dear @izabala123 

Now it can export smoothly. But I found another problem.  When I import my case of total mesh including both hull mesh and waterplane mesh. The summary shows that it has 1487 panels. But it acutally should have 1488 panels. 

Below is my mesh. Mesh_1.dat is the salome mesh including both hull mesh and waterplane mesh. HullMesh.pnl and WaterplaneMesh.pnl are hull mesh and waterplane mesh splitted from Mesh_1.dat respectively by my own fortran code. You can see that waterplane has 67 panels. And Hull mesh has 1421 panels. 
[mesh.zip](https://github.com/BEMRosetta/BEMRosetta/files/11008625/mesh.zip)

If I use "Get Hull" and "Get Waterplane" functionality in "process"  of BEMRosetta to split the salome mesh. The hull is correct. But the "Waterplane" lacked one panel, as shown below:

![image](https://user-images.githubusercontent.com/56197603/226109085-a255d12a-699e-4b05-b0ec-2076e13e50c6.png)

So I assume when the total panels are imported in to BEMRosetta, one panel is lacked.

Regards,

**Comment by izabala123** [comment]
Hello Ran

In the original Nemoh.dat I see that this panel is not there. You can see
that in the picture.
[image: image.png]

Doing first Get Hull, and later Fill Waterplane, you can get this:

[image: image.png][image: image.png]

Anyway I hope to improve Fill Waterplane in the next month, solving some
problems with special shapes like multiple concave water planes.

Best regards
Iñaki

**Comment by Ran-Tu** [comment]
Dear @izabala123 

Yes, the original Nemoh.dat only includes hull mesh, and waterplane mesh is excluded. It is a good approach to only use the hull mesh and create the waterplane mesh by "Fill Waterplane".

But for the newly uploaded folder, it is a mesh including both hull and waterplane. In this condition, I used Get Hull and Get Waterplane, and I hadn't used Fill Waterplane. 

I found that BEMRosetta only read 1487 panels while it acutally should have 1488 panels. And I found the lost panel is on the waterplane, by comparing the BEMRosetta exported file to my fortran-exported file.  In the newly uploaded folder, Mesh_1.dat is the salome mesh including both hull mesh and waterplane mesh. HullMesh.pnl and WaterplaneMesh.pnl are generated by my fortran code. I have confirmed that the number of panels in my fortran-exported file can match Salome inquiry results.

 I am concerned that when BEMRosetta read a mesh including both hull and waterplane, it possibly neglect one row of file so that one panel is lost.

Regards,

**Comment by izabala123** [comment]
Dear Ran

I see that answering you by email, the images have been lost, so I will repeat the message.

In the file mesh.zip you have sent me today, there is a file Nemoh.dat, In this file, the panel you miss is not there. This way, BEMRosetta cannot get it, as it doesn't exist. You can see that in the picture.

![image](https://user-images.githubusercontent.com/38589221/226129320-968ae6a9-c613-414b-87fd-69414dc0697c.png)

Doing first Get Hull, and later Fill Waterplane, you can get this:

![image](https://user-images.githubusercontent.com/38589221/226129354-0bbdcb95-db77-4507-9d9d-f948a671ff45.png)
![image](https://user-images.githubusercontent.com/38589221/226129366-cc3da9e9-9846-4877-bae4-c102c5b763e7.png)

Anyway I hope to improve Fill Waterplane in the next month, solving some
problems with special shapes like multiple concave water planes.

Best regards
Iñaki

**Comment by Ran-Tu** [comment]
Dear @izabala123 

 I created a simple case to explain it. 

1. In Salome, I created a cube, and the each face of the cube is discretized into 4 quadrilateral panels. So this cube should totally have 26 nodes, 24  and quadrilateral panels. And I can prove it by inquiring in Salome (as shown in below figure):

![Screenshot 2023-03-19 100052](https://user-images.githubusercontent.com/56197603/226149612-212aba89-35fe-4561-ba62-8a7de1f1e66e.png)

2. Then I export cube.dat file. As you can see in below figure, the index of type 204 element is from 25 to 48. So we can also conclude that cube.dat file contains ( 48 - 25 ) +1 = 24 quadrilateral panels.

![image](https://user-images.githubusercontent.com/56197603/226149855-2d6a407a-42e2-42e5-b5ad-8f73361cdfd6.png)

Now I can confirm cube.dat file does not exit any problem as it has the same number (24) of quadrilateral panels with Salome inquiry result.

3. Finally I import cube.dat into BEMRosetta. But BEMRosetta only reads 23 panels. One panel is missed.  

![image](https://user-images.githubusercontent.com/56197603/226149816-a25d5fd4-f721-4acc-8376-2042d8869630.png)

What I want to say is that my cube mesh file indeed has 24 panels, but BEMRosetta only finds 23 panels in my cube mesh file, and one panel is ingored somehow by BEMRosetta. I assume there is a bug in BEMRosetta to miss one panel.

My cube mesh file is uploaded below:
[cube.zip](https://github.com/BEMRosetta/BEMRosetta/files/11010037/cube.zip)

Regards,

**Comment by izabala123** [comment]
Oh, now I understand. 

Thank you for your super simple sample. With it, solving the problem has been very easy.
In few hours a fixed version will be uploaded.

Best regards
Iñaki

**Comment by Ran-Tu** [comment]
Dear @izabala123 

Now the fixed version can find the total 24 panels and 26 nodes for this cube mesh!

But when I use Get Hull to extract hull mesh from the cube, I find that BEMRosetta shows that the extracted hull mesh has 26 nodes (as shown below).
![Screenshot 2023-03-24 101157](https://user-images.githubusercontent.com/56197603/227406819-44dc9d07-574d-4fce-a710-af2eb7e410df.png)

**But there is an unexpected node at the middle of water surface, as you can see below.** This unexpected node does not construct any panel.
![Screenshot 2023-03-24 101701](https://user-images.githubusercontent.com/56197603/227407569-94622893-10fa-43b4-a824-1d0a4a7f8101.png)

That is, the extracted hull mesh should only have 25 nodes. However, now BEMRosetta finds 26 nodes when extracting hull from cube. An unexpected node was not removed from the hull mesh during Get Hull process.

Regards,

**Comment by izabala123** [comment]
Hi Ran-Tu

Thank you for your interest and for your attention to detail. It is true that initially this was done on purpose, and that it can be removed with the Process/Simplify option. However, looking at it now, I think the best thing to do is what you say. So, in a few hours I'll upload a version that cleans up the leftover elements after doing this operation.

Best regards
Iñaki

**Comment by Ran-Tu** [comment]
Dear @izabala123 

Oh I had not known the functionality of Process/Simplify option before I saw your post. After using Simplify, BEMRosetta remove all unexpected nodes. Now everything is OK. I can smoothly convert Salome dat mesh to HAMS pnl mesh, and then conduct simulation. It is great! 

It is ok for current version. And it would be better if Get Hull can automaically remove the unnecessay nodes which construct no panels.

By the way, is there any user manual for BEMRosetta to help users better understand each options?

Regards,
Ran
